overlord/snapstate/storehelpers.go

```diff
@@ -82,6 +82,101 @@ func refreshOptions(st *state.State, origOpts *store.RefreshOptions) (*store.Ref
 	return &opts, nil
 }
 
+// installSizeInfo returns total download size of snaps and their prerequisites
+// (bases and default content providers), querying the store as neccessarry,
+// potentially more than once. It assumes the initial list of snaps already has
+// download infos set.
+// The state must be locked by the caller.
+func installSizeInfo(st *state.State, snaps []*snap.Info, userID int) (uint64, error) {
+	curSnaps, err := currentSnaps(st)
+	if err != nil {
+		return 0, err
+	}
+
+	user, err := userFromUserID(st, userID)
+	if err != nil {
+		return 0, err
+	}
+
+	skipSnaps := map[string]bool{}
+	for _, snap := range curSnaps {
+		skipSnaps[snap.InstanceName] = true
+	}
+
+	var prereqs []string
+
+	resolveBaseAndContentProviders := func(snapInfo *snap.Info) {
+		if snapInfo.SnapType != snap.TypeApp {
+			return
+		}
+		if snapInfo.Base != "none" {
+			base := defaultCoreSnapName
+			if snapInfo.Base != "" {
+				base = snapInfo.Base
+			}
+			if !skipSnaps[base] {
+				prereqs = append(prereqs, base)
+				skipSnaps[base] = true
+			}
+		}
+		for _, snapName := range defaultContentPlugProviders(st, snapInfo) {
+			if !skipSnaps[snapName] {
+				prereqs = append(prereqs, snapName)
+				skipSnaps[snapName] = true
+			}
+		}
+	}
+
+	var total uint64
+	for _, snapInfo := range snaps {
+		if snapInfo.DownloadInfo.Size == 0 {
+			// XXX: we could support this by simply adding to prereqs
+			return 0, fmt.Errorf("internal error: download info missing for %q", snapInfo.InstanceName())
+		}
+		total += uint64(snapInfo.Size)
+		resolveBaseAndContentProviders(snapInfo)
+	}
+
+	opts, err := refreshOptions(st, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	theStore := Store(st, nil)
+	// calls to the store should be done without holding the state lock
+	st.Unlock()
+	defer st.Lock()
```

Comment:
once we unlock other changes can happen that touch the same snaps/prereqs,  the full solution to that is complicated as we discussed as it requires caching and attaching size information to changes,  but we can maybe start at least structuring this so it's closer to what we'll need for that at some point;  instead of computing the total, I recommend just keeping a map of snaps -> size  then after the loop require the lock and re-acquire the list of current snaps and adding only those that are not in the new current list...

Reply:
Done.